Modules/gcmodule.c

```diff
@@ -2348,13 +2348,14 @@ PyVarObject *
 _PyObject_GC_Resize(PyVarObject *op, Py_ssize_t nitems)
 {
     const size_t basicsize = _PyObject_VAR_SIZE(Py_TYPE(op), nitems);
+    const size_t presize = _PyType_PreHeaderSize(((PyObject *)op)->ob_type);
     _PyObject_ASSERT((PyObject *)op, !_PyObject_GC_IS_TRACKED(op));
-    if (basicsize > (size_t)PY_SSIZE_T_MAX - sizeof(PyGC_Head)) {
+    if (basicsize > (size_t)PY_SSIZE_T_MAX - presize) {
         return (PyVarObject *)PyErr_NoMemory();
     }
 
     PyGC_Head *g = AS_GC(op);
-    g = (PyGC_Head *)PyObject_Realloc(g,  sizeof(PyGC_Head) + basicsize);
+    g = (PyGC_Head *)PyObject_Realloc(g,  presize + basicsize);
```

Comment:
These lines (2357-2361) will not be correct if `presize` includes a managed dict. The layout would look like:

```
[ weak ptr ] [ managed dict ] [ _gc_prev ] [ _gc_next ] [PyObject ....]
^alloc                        ^ g                       ^ op
```

The bugs (when including manage dict/weak ptr) are:
1. You can't call PyObject_Realloc on `g`, it needs to be called on what I labelled as `alloc` in my diagram above.
2. The returned value from `PyObject_Realloc` will be the beginning of the allocation (not necessarily the PyGC_Head)

Reply:
Ah thanks for the pointing out. I will take a look.

Reply:
@colesbury 
I've updated the PR. Thanks for pointing it out; this is my first patch related to GC.
I wish that my changes handle the object header properly.